hooks/src/index.js

```diff
@@ -132,19 +140,16 @@ export function useRef(initialValue) {
 	return useMemo(() => ({ current: initialValue }), []);
 }
 
+/**
+ * @param {object} ref
+ * @param {() => object} createHandle
+ * @param {any[]} args
+ */
 export function useImperativeHandle(ref, createHandle, args) {
-	const state = getHookState(currentIndex++);
-	if (argsChanged(state._args, args)) {
-		state._args = args;
-		currentComponent.__hooks._handles.push({ ref, createHandle });
-	}
-}
-
-function bindHandles(handles) {
-	handles.some(handle => {
-		if (handle.ref) handle.ref.current = handle.createHandle();
-	});
-	return [];
+	args.push(ref);
```

Comment:
Not sure how our general stance on this topic is but we are modifying an array from user-land here. This might have strange side effects for users. What do you think about this?

Reply:
I’ve thought about this and this introduces non-deterministic behavior what if the user sends no array (happens on render) crash/limitation

Reply:
Oh good points. I'll fix up and add tests

Reply:
Suggestion

```suggestion
	args = args.concat(ref);
```

Reply:
Problem with that is that this can diminish the on every render to become once and never again

Reply:
Hmm - I'm not sure what you mean. My understanding is that `useImperativeHandle()` is always going to get called on every render (even if it returns a cached value). The `args` value will always be new unless someone hoists it out of render, in which case modifying it in-place will cause an infinite loop.

Reply:
Yes but the args array is optional meaning that it can't cache at all. If we always make it an array it is dependent on a mutable ref which potentially does not change. This makes it so that useImperativeHandle(() => ) is equivalent to useImperativeHandle(() => , []) which is not deterministic

Reply:
Updated this code so that `useImperativeHandle` should support all the functionality of `useLayoutEffect` and added to tests to capture missing features I think

Reply:
For anyone curious, [here is the relevant React code](https://git.io/Je8mb) I believe